src/api/blob/controller.go

```diff
@@ -188,6 +160,69 @@ func (c *controller) Exist(ctx context.Context, digest string, options ...Option
 	return true, nil
 }
 
+func (c *controller) FindMissingAssociationsForProject(ctx context.Context, projectID int64, blobs []*blob.Blob) ([]*blob.Blob, error) {
+	if len(blobs) == 0 {
+		return nil, nil
+	}
+
+	var digests []string
+	for _, blob := range blobs {
+		digests = append(digests, blob.Digest)
+	}
+
+	associatedBlobs, err := c.blobMgr.List(ctx, blob.ListParams{BlobDigests: digests, ProjectID: projectID})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(associatedBlobs) == 0 {
+		return blobs, nil
+	} else if len(associatedBlobs) == len(blobs) {
+		return nil, nil
+	}
+
+	associated := make(map[int64]bool, len(associatedBlobs))
+	for _, blob := range associatedBlobs {
+		associated[blob.ID] = true
+	}
+
+	var results []*models.Blob
+	for _, blob := range blobs {
+		if !associated[blob.ID] {
+			results = append(results, blob)
+		}
+	}
+
+	return results, nil
+}
+
+func (c *controller) Get(ctx context.Context, digest string, options ...Option) (*blob.Blob, error) {
```

Comment:
if I only pass the digest, what blob should I get?

Reply:
The blob of the digest itself.